### DISCUSSION TITLE
Insert block with attributes

### DISCUSSION THREAD

**Discussion by w4l1k** [DomCR/ACadSharp#444] [Q&A] [answered]
Hi,
how can I insert a block with attributes into the model? 
I can't figure out how to do it, here is my code .
I only see the circle on the model without the text.
When I insert the block manually, the text appears.

		CadDocument doc = new(ACadVersion.AC1024);
		BlockRecord record = new("my_block");

		record.Entities.Add(new Circle
		{
			Radius = 10,
			Center = XYZ.Zero
		});

		record.Entities.Add(new AttributeDefinition()
		{
			InsertPoint = XYZ.Zero,
			Prompt = "Name ???",
			Tag = "CIRCLE_NAME",
			Value = "Circilla",
			HorizontalAlignment = TextHorizontalAlignment.Left,
			Height = 18,
			AttributeType = AttributeType.SingleLine,
		});

		doc.BlockRecords.Add(record);

		doc.Entities.Add(new Insert(record)
		{
			InsertPoint = new XYZ(100, 100, 0),
			XScale = 0.8,
			YScale = 0.8,
		});

		using DwgWriter writer = new("../../../test.dwg", doc);
		writer.OnNotification += NotificationHelper.LogConsoleNotification;
		writer.Write();

**Comment by DomCR** [comment, marked as answer]
Hi @w4l1k,

Attributes should be added into `Insert.Attributes`, you will notice that when you create the `Insert` referencing the `BlockRecord` the collection should contain one `AttributeEntity` in it, to change the value, fetch the `AttributeEntity` and edit the object.

This is a feature that should be improved with this PR:

- https://github.com/DomCR/ACadSharp/pull/446

With your changes the code should work fine, if you save a DXF the file should be as expected, for the DWG I forgot to remove a guard that blocks the writer for Attributes, I'll open an issue to fix it.

If you are using the package, please let me know and I'll make an update.

Let me know if this helps.

**Reply by w4l1k** [reply]
Thank you for your help.
I'm using the 2.4.0-beta package from NuGet.
It works well in DXF, but unfortunately, I  (using/need) DWG.


		CadDocument doc = new(ACadVersion.AC1024);
		BlockRecord record = new("my_block");

		record.Entities.Add(new Circle
		{
			Radius = 10,
			Center = XYZ.Zero
		});

		record.Entities.Add(new AttributeDefinition()
		{
			InsertPoint = XYZ.Zero,
			Prompt = "Name ???",
			Tag = "CIRCLE_NAME",
			Value = "Circilla",
			HorizontalAlignment = TextHorizontalAlignment.Left,
			Height = 18,
			AttributeType = AttributeType.SingleLine,
		});

		doc.BlockRecords.Add(record);

		Insert insert = (new Insert(record)
		{
			InsertPoint = new XYZ(100, 100, 0),
			XScale = 0.8,
			YScale = 0.8,
		});

		insert.Attributes.First(a => a.Tag == "CIRCLE_NAME").Value = "It Works !!!";

		doc.Entities.Add(insert);

		using DwgWriter writer = new("../../../test.dwg", doc);
		using DxfWriter dxfWriter = new DxfWriter("../../../test.dxf", doc, false);
		writer.OnNotification += NotificationHelper.LogConsoleNotification;
		dxfWriter.OnNotification += NotificationHelper.LogConsoleNotification;
		writer.Write();
		dxfWriter.Write();

**Reply by DomCR** [reply]
I've patched master, I'll make a release shortly.

FYI, the package has a v1.0.0 release.

**Reply by w4l1k** [reply]
One more question, when I do an INSERT the attributes value do not follow the block.
circle (100,100,0) attribute value (0,0,0).
After running the ATTSYNC command, the attributes value are in the correct positions.
When I insert the block manually, it's fine"

**Reply by DomCR** [reply]
With the last patch it seems to work fine, if this happens after the release, please let me know and I'll take a look.

**Reply by DJGosnell** [reply]
> One more question, when I do an INSERT the attributes value do not follow the block. circle (100,100,0) attribute value (0,0,0). After running the ATTSYNC command, the attributes value are in the correct positions. When I insert the block manually, it's fine"

When looking into this in the past, the attribute locations are absolute.  The only reason the attributes are updated in AutoCAD is because AutoCAD moves all the attributes relatively to the insert.  It is an extra step that needs to be considered and implemented by the user.
